### ISSUE TITLE
Kopernicus support

### ISSUE THREAD

**Issue by Gordon-Dry** [Kerbalism/Kerbalism#87]
You know the deal :)

**Comment by Gordon-Dry** [comment]
The planner then also needs to show all first node child bodies of a "star", which are moons in reality. So that these "planets" are also available as reference bodies in the planner. Incl. the check if gas or rock, so if surface is possible.

**Comment by gotmachine** [comment]
- Multiple suns/stars support is probably never gonna happen, as this affect too many aspects of the background simulation we are doing. We already have major unresolved issues while dealing with only 1 sun for calculating sunlight, temperature, radiation for all vessels at once.
- Systems with only 1 sun, but that do sun reparenting could be supported, this doesn't require too much work.

**Comment by Tonas1997** [comment]
`Multiple suns/stars support is probably never gonna happen, as this affect too many aspects of the background simulation we are doing. We already have major unresolved issues while dealing with only 1 sun for calculating sunlight, temperature, radiation for all vessels at once.`

I assume this includes star systems that have the Sun as a parent body, correct?

**Comment by SirMortimer** [comment]
yeah, that didn't age very well. @gotmachine kind of changed his mind since then :)

**Comment by gotmachine** [comment]
Yup, full Kopernicus stars support is available in the sstu_and_solar dev branch. See https://github.com/Kerbalism/Kerbalism/commit/1f108c1a0bb305f44291e366a5ce09f52da57946 for details.

This said, my statement still stands. The support code is already quite consequent and should be enough in most cases for our current needs. But : 
- There are a lot of small corner-cases inconsistencies that I had to let go for the sake of keeping this sane. 
- The current issues we have with high-warp evaluation will be magnified in multi-stars configurations
- The envisioned (in project) fix for this issue may (or not) cause performance issues with multiple stars.

**Comment by Tonas1997** [comment]
Thanks for the answer! I commented because I'm having an issue wherein distant stars (>4 ly) charge local solar panels at ridiculous rates; while it is probably related to some Kopernicus config, it could've also been related to some Kerbalism incompatibility.

**Comment by Tonas1997** [comment]
Just read the dev branch nodes and goddamn, it will be one heck of a livesaver 😄 

I also posted about my issue on Kopernicus' GitHub page. For the sake of closure, are instantaneous (e.g. as oppose to timewarp) solar power calculations handled in any way by Kerbalism? If yes, my problem would have nothing to do with Kopernicus and I won't need to bother those good folks any further...

**Comment by gotmachine** [comment]
If you're using the kerbalism dev version (sstu_and_solar branch), yes all solar panel calculations are overridden by Kerbalism, not matter the situation.

**Comment by Tonas1997** [comment]
I'm using the release version 3.0.2, so I guess my problem has nothing to do with Kerbalism. My installed version only manages charge rates during timewarp, right?

**Comment by gotmachine** [comment]
3.0.2 is also overriding all calculations. But it won't work correctly if you use a Kopernicus-modified system with multiple stars (or reparented sun). This is the point of this issue and of the new code in the dev branch.

**Comment by Tonas1997** [comment]
Sorry, I thought Kerbalism was only handling background resources.

So I just need to grab the GameData/Kerbalism folder from the dev branch and replace mine with it? And what should I do with the latest binaries you linked?

**Comment by gotmachine** [comment]
Use only the GameData/Kerbalism folder from the link above

**Comment by Tonas1997** [comment]
Thanks for the help. I'll try it and report back.

**Comment by Tonas1997** [comment]
Reporting back: sadly, the problem persists on` sstu-and-solar`. However, I made some more tests to help and nail down the problem:

![screenshot34](https://user-images.githubusercontent.com/43859708/62504189-438ea780-b7ef-11e9-8ebd-a59f3ce92f8e.png)

Here, the probe is orbiting Proxima Centauri B and only has direct line-of-sight with the Sun. Charging at 0.45 EC/sec.

![screenshot36](https://user-images.githubusercontent.com/43859708/62504218-68831a80-b7ef-11e9-9a8c-08892c328ca1.png)

![screenshot35](https://user-images.githubusercontent.com/43859708/62504221-6de06500-b7ef-11e9-8b0e-6bf177021523.png)

In these images, the probe has no LOS with any star whatsoever and is consuming electric charge at a nominal rate.

**TL;DR:** direct line of sight with a star will always generate EC at a fixed rate of ~0.40 EC/sec., regardless of distance or panel orientation.*

*as a side note, the panel was always turned towards the the Sun.

**Comment by gotmachine** [comment]
Sorry, didn't see that you are using custom configs. 
Old configs won't work with the dev branch as there is a new partmodule that need to be added to solar panels parts. 

See docs here : https://github.com/Kerbalism/Kerbalism/wiki/TechGuide-~-PartModules-~-SolarPanelFixer

If it's working, you should see the following PAW UI in the editor : 
![image](https://user-images.githubusercontent.com/24925209/62504814-8609b180-b7fa-11e9-9120-88998c7a079c.png)

And in flight :
![image](https://user-images.githubusercontent.com/24925209/62504949-f31d4700-b7fa-11e9-8e59-dfd0c408b81d.png)

**Comment by Tonas1997** [comment]
Can't I just add the new module to all solar panels in a makeshift patch (to avoid switching configs)?

**Comment by UsernamesAreAlwaysTaken** [comment]
You can. There was something mentioned somewhere else that the solar panel fixer module has to be after the actual panel module, so your patch must be running pretty late.

**Comment by Gordon-Dry** [comment]
Like in (ROKerbalism's) GameData\KerbalismConfig\System\Misc.cfg where it says
```
@PART[*]:HAS[@MODULE[ModuleDeployableSolarPanel]]:FOR[Kerbalism]
{
  MODULE
  {
    name = WarpFixer
  }
} 

```
perhaps FOR -> AFTER and Kerbalism -> zzzKerbalism ?

**Comment by Gordon-Dry** [comment]
But still this is the question:
FOR -> AFTER | Kerbalism -> zzzKerbalism ?

**Comment by gotmachine** [comment]
`[zzzKerbalism]` isn't needed in this case and there are other patches that depend on it to be in `[Kerbalism]`. 

The only rule to follow is that if you have a patch that add a solar panel module to the part (very unlikely), it must be done in `:BEFORE[Kerbalism]`.
See the MOLE support config for an example.

**All the the config changes :**

- Main patch in KerbalismConfig (Misc.cfg) : 
https://github.com/Kerbalism/Kerbalism/blob/a21905901110900fe9fafa4b8da8c343e53a2700/GameData/KerbalismConfig/System/Misc.cfg#L9

- Reliability patch (Reliability.cfg) : 
https://github.com/Kerbalism/Kerbalism/blob/a21905901110900fe9fafa4b8da8c343e53a2700/GameData/KerbalismConfig/System/Reliability.cfg#L59

Mod support patches. Those are currently in the core Kerbalism folder, but we are in the process of moving them to KerbalismConfig.

- Kopernicus : 
https://github.com/Kerbalism/Kerbalism/blob/sstu-and-solar/GameData/Kerbalism/Support/Kopernicus.cfg

- SSTU :
https://github.com/Kerbalism/Kerbalism/blob/a21905901110900fe9fafa4b8da8c343e53a2700/GameData/Kerbalism/Support/SSTU.cfg#L411

- Near future solar :
https://github.com/Kerbalism/Kerbalism/blob/sstu-and-solar/GameData/Kerbalism/Support/NearFuture.cfg#L146

- MOLE (just changed the patch order to BEFORE)
https://github.com/Kerbalism/Kerbalism/blob/1edb8a6cab0b67369e01f18b85622b70f1697863/GameData/Kerbalism/Support/MOLE.cfg#L56

**Comment by Gordon-Dry** [comment]
@Standecco ^

**Comment by Gordon-Dry** [comment]
@gotmachine  first line of that patch should be PART[*]:HAS ...

How about this?
```
@PART[*]:HAS[@MODULE[ModuleDeployableSolarPanel]]:NEEDS[RealismOverhaul,Kerbalism]:FOR[zzzKerbalism]
{
	@MODULE[ModuleDeployableSolarPanel]
	{
		%useKopernicusSolarPanels = false
	}
}

@PART[*]:HAS[@MODULE[ModuleROSolarPanel]]:NEEDS[RealismOverhaul,Kerbalism]:FINAL
{
  !MODULE[ModuleROSolarPanel] {}
}

@PART[*]:HAS[@MODULE[WarpFixer]]:NEEDS[RealismOverhaul,Kerbalism]:FINAL
{
  !MODULE[WarpFixer] {}

  MODULE
  {
    name = SolarPanelFixer
  }
}
```

**Comment by Tonas1997** [comment]
A few notes after testing, now with the correct patches and modules:

- The panel charge rates simulation on the VAB is absolute bonkers... which isn't a GUI problem, as I'll show later.

![screenshot48](https://user-images.githubusercontent.com/43859708/62641969-a2ae0280-b93c-11e9-915e-8b82ce0446b5.png)

- Panels still charge at a considerable rate when aiming at distant stars. Also, auto-aim function doesn't select the nearest/highest output star.

![screenshot57](https://user-images.githubusercontent.com/43859708/62642225-27008580-b93d-11e9-947a-11eb97bddac1.png)

![screenshot60](https://user-images.githubusercontent.com/43859708/62642187-1819d300-b93d-11e9-8823-db9837edbcc1.png)

- Solar input near "other" stars is through the roof:

![screenshot52](https://user-images.githubusercontent.com/43859708/62642294-48fa0800-b93d-11e9-911d-feafd52dfd0d.png)

Additionally,

- What's the difference between solar power, exposure and EC/s? There seems to be a discrepancy between the latter and *actual* charge rate, as seen on the resource panel and the PAW:

![screenshot60](https://user-images.githubusercontent.com/43859708/62642404-8b234980-b93d-11e9-8efd-7da09ce917e4.png)

- Since you're working with Kopernicus, do you think this could be happening due to misconfigured stars? I seem to remember a `SolarPowerCurve` module, which has since been deprecated - nevertheless, the Sun itself lacks that module and works fine.

Thanks for the patience and all the help - past and future - during these trying times 😃

**Comment by gotmachine** [comment]
Well the "solar power" information is not from Kerbalism. I suspect there is another mod interacting in an odd way, but there is probably some bugs on our side, this stuff has been very little tested.

I can't really know just at looking at screenshots. Could you upload : 
- your KSP.log file
- your ModuleManager.ConfigCache file (inside gamedata, zip it first, it's huge otherwise)
And give me the list of planet packs you are using.

**Comment by Tonas1997** [comment]
Here's the ZIP file with both KSP.log and ModuleManager cache files.

- The log was generated over a _generously_ long playthrough. The tests were made on savefile "Testes Simples" (from the start of the log up to line 183308)

- The planet pack was made for KSP version 1.3.1* and has since become unavailable; I'm not sure if sharing the whole pack would count as piracy. If not (and in the case you really need the configs), I could share the folder privately. Either way, the bodies can be found on the MM cache.

*since they otherwise work fine in 1.6.1 and there hasn't been any major syntax changes since, I assume the pack is working as intended in more recent versions of Kopernicus. 

[Documents.zip](https://github.com/Kerbalism/Kerbalism/files/3478770/Documents.zip)

**Comment by gotmachine** [comment]
1. There was a mistake in how I was evaluating Kopernicus stars luminosity, that's what causing the huge ec rates. It's fixed but not available yet (see https://github.com/Kerbalism/Kerbalism/commit/6a26a1c7151cf80fb9c9de371fcaea0fc8a17518)

2. The only thing we use to determine the brightness / energy flux is the `Body > ScaledVersion > Light > luminosity` value. This is the solar flux in W/m² at a distance of 1AU from the star, where 1AU is the distance between the home body and the home body parent star. 

All other luminosity values/curves are ignored, this is also how Kopernicus evaluate the flux in it's own solar panel module. I'm prettu sure those other values/curves are only used for graphic/rendering purposes.

**Comment by Tonas1997** [comment]
So I presume that `luminosity` value and the vessel's distance to the star is used to calculate the panel energy flow.

P.S.: my problem was being caused by KSPI-E, which has since been patched 😃

**Comment by gotmachine** [comment]
Yup. For reference, our sun luminosity at sun-earth distance (1 "real" AU) is about 1360 W/m², which is also the value used for kerbin's sun at sun-kerbin distance (1 "KSP" AU).

**Comment by Tonas1997** [comment]
Concerning Kerbalism/SSTU support, is the current indev work covering the issues I reported in #387 ?

**Comment by gotmachine** [comment]
Yes. Please don't spam this issue with unrelated comments. For this kind of questions, please ask on our [discord server](https://discord.gg/3JAE2JE)

**Comment by Shiolle** [comment]
This is a Kerbalism issue. There is a logical error in the SunData constructor: it treats the distance between the home body and the star for which the object being created as a measure of that stat's AU, which results in inflated solarFluxTotal values. I will create a pull request with a possible solution.

**Comment by gotmachine** [comment]
@Shiolle This is fixed in the dev branch. Read a few comments above.

**Comment by Shiolle** [comment]
Sorry.
